File: src/algorithms/cpp/bst.cpp
// Binary Search Tree Insertion
#include <iostream>
#include <vector>
#include <string>

#define EMIT(type, id) std::cout << "{\"type\": \"" << type << "\", \"nodeIds\": [\"" << id << "\"]}\n"

using namespace std;

struct Node {
    int key;
    std::string id;
    Node* left;
    Node* right;
    Node(int item) {
        key = item;
        id = "node-" + std::to_string(item);
        left = right = nullptr;
    }
};

Node* insert(Node* node, int key) {
    if (node == nullptr) {
        EMIT("INSERT", "node-" + std::to_string(key));
        return new Node(key);
    }
    
    EMIT("COMPARE", node->id);
    if (key < node->key)
        node->left = insert(node->left, key);
    else if (key > node->key)
        node->right = insert(node->right, key);
        
    return node;
}

int main() {
    vector<int> arr = {15, 10, 20, 8, 12, 17, 25};
    Node* root = nullptr;
    
    for (int val : arr) {
        root = insert(root, val);
    }
    
    return 0;
}

File: src/algorithms/python/bst.py
# Binary Search Tree Insertion
import json
import time


def emit(event_type, **kwargs):
    event = {"type": event_type}
    event.update(kwargs)
    print(json.dumps(event))


class Node:
    def __init__(self, key):
        self.key = key
        self.id = f"node-{key}"
        self.left = None
        self.right = None


def insert(root, key):
    if root is None:
        emit("INSERT", nodeIds=[f"node-{key}"])
        return Node(key)
    else:
        emit("COMPARE", nodeIds=[root.id])
        if root.key < key:
            root.right = insert(root.right, key)
        else:
            root.left = insert(root.left, key)
    return root


# Example usage:
arr = [15, 10, 20, 8, 12, 17, 25]
root = None

emit("SYSTEM_LOG", level="INFO", message="Building BST...")
for val in arr:
    root = insert(root, val)
    time.sleep(0.05)
